src/scheduler/batch/tasks/populateMissingAnnualReviewDates/__tests__/startDateFromFirstPublishedDate.test.ts

```diff
@@ -0,0 +1,37 @@
+import { startDateFromFirstPublishedDate } from "../startDateFromFirstPublishedDate";
+
+test("returns a date 1 year in the future when firstPublishedDate is upto 11 months in the past", () => {
```

Comment:
I don't know if the  `startDateFromFirstPublishedDate` functions needs this many unit tests. You're welcome to keep them but from looking at the function, the main logic is being handled by `date-fns` which should already be tested.

The logic we've implemented is this ternary  `startDate = proposedDateIsWithinAMonth ? minDate : proposedDate;`, which at mosts needs 2-3 unit tests imo. But it would be good to hear your thoughts.

Reply:
I agree that you don't need to test what's already tested by a library. And I agree that the logic we're testing is essentially the ternary. I am doing 2-3 unit tests based on the logic though (but multiple expects per test). If one of these expects fail - the unit test fails, and therefore the business logic has been broken. 


So I need to test at least two branches of code `proposedDateIsWithinAMonth`. They're tested by `returns a date 1 year in the future when firstPublishedDate is upto 11 months in the past` and `returns a date 29 days from now`. 

although it looks like I'm testing if the date-fns library is working -- I'm actually testing the implementation of the logic. 

For `returns a date 1 year in the future when firstPublishedDate is upto 11 months in the past` - I check 4 dates in that 11 month range, from the beginning, middle and end of the ranges. I also picked a system time where 11 months ago would be this year, and the previous year. I vaguely remember some code was implemented in the past (I think around annual review date), which just checked if the current year is correct. So what if this function was implemented in a similar way? what if the code is just checking today's date + 1 year? Even though 2022-04-01 is valid, it would fail because the start date is not 2024. 


for `returns a valid date when firstPublished was on a leap day`, is essentially testing another library, but it gives us confidence we know what to expect from the library, i.e. how does it handle leap days. does it +1 into the next month (1 march)? or use 28th feb? 

`returns a date 29 days from now` tests dates in the past, and dates in the future (but only up to 29 days in the future). 

`when firstPublished was over 11 months ago` - I can probably get away with just testing 2 dates in the past. One just before the window, and one way over the window. 

`when firstPublished was between 11 and 12 months ago` - Similar to the first unit test. I test a few dates within the window, the start/ends of the window and the dates between.  

Having 1 expect per test won't be sufficient. although your line coverage might be high, it doesn't give you confidence. I've written the code in a way where it's very easy and fast to test multiple variations, instead of having to mock prisma or rely on a database. 

For the most part I try to ignore implementation detail. I don't need to spy on `isBefore` to check if it's doing the right thing. I just want to check what comes out is correct and fits with business logic.